libafl/src/mutators/string.rs

```diff
@@ -0,0 +1,533 @@
+//! Mutators for preserving string categories, which may be useful for certain targets which are primarily string-oriented.
+use alloc::vec::Vec;
+use core::{cmp::Ordering, ops::Range};
+
+use libafl_bolts::{rands::Rand, Error, HasLen, Named};
+
+use crate::{
+    inputs::{BytesInput, HasBytesVec},
+    mutators::{rand_range, MutationResult, Mutator, Tokens},
+    state::{HasMaxSize, HasMetadata, HasRand},
+};
+
+/// Unicode category data, as used by string analysis and mutators.
+pub mod unicode_categories {
+    #![allow(unused)]
+    #![allow(missing_docs)]
+
+    include!(concat!(env!("OUT_DIR"), "/unicode_categories.rs"));
+}
+
+/// Mutator which retains the general category of a randomly selected range of bytes
+#[derive(Debug, Default)]
+pub struct StringCategoryPreservingMutator;
+
+impl Named for StringCategoryPreservingMutator {
+    fn name(&self) -> &str {
+        "string-category-preserving"
+    }
+}
+
+const MAX_CHARS: usize = 16;
+
+fn choose_index<R: Rand>(rand: &mut R, bytes: &[u8]) -> Option<(char, usize)> {
+    if bytes.is_empty() {
+        None
+    } else {
+        let idx = rand.below(bytes.len() as u64);
+        let mut indices = (idx.saturating_sub(3)..=idx).rev();
+
+        // try to find a character between [idx-3, idx+4)
+        let (c, idx, max_idx) = 'outerloop: loop {
+            if let Some(idx) = indices.next() {
+                let idx = idx as usize;
+                for max_idx in (idx + 1)..=(idx + 3).min(bytes.len()) {
```

Comment:
maybe I'm just wrong
should we add .rev() here to search from longer utf chars?

the current code looks like it is finding shortest valid utf chars

Reply:
I think we should aim to prefer shorter UTF-8? That represents more common use cases. Potentially, we could select over a set of valid options?